service/worker/scheduler/workflow.go

```diff
@@ -203,7 +211,18 @@ func (s *scheduler) run() error {
 	s.pendingPatch = s.InitialPatch
 	s.InitialPatch = nil
 
+	// initialize the generator
+	s.nextTimeGenerator = s.timeGenerator()
+
+	currentVersion := s.tweakables.Version
+	currentSpec := s.cspec
+
 	for iters := s.tweakables.IterationsBeforeContinueAsNew; iters > 0 || s.pendingUpdate != nil || s.pendingPatch != nil; iters-- {
+		// if spec changes or version is changed (donwgrade or upgrade), break and continue-as-new
+		// to force invalidation of internal caches
+		if currentVersion != s.tweakables.Version && currentSpec != s.cspec {
```

Comment:
I think this is too harsh. I was thinking more like: if spec changes, just invalidate cache. If version changes... well, in most cases I think CAN is unnecessary. But maybe we should for downgrades? Let's discuss this

Reply:
Also, for spec, you can do the invalidation in `compileSpec`, since all changes must go through there